File: mmupnp/src/test/java/net/mm2d/upnp/MockMulticastSocket.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.InetAddress;
import java.net.MulticastSocket;
import java.net.NetworkInterface;
import java.net.SocketException;

public class MockMulticastSocket extends MulticastSocket {
    private DatagramPacket mDatagramPacket;
    private InetAddress mInetAddress;
    private byte[] mReceiveData;
    private long mWait;

    public MockMulticastSocket() throws IOException {
        super();
    }

    @Override
    public void setTimeToLive(final int ttl) {
    }

    @Override
    public void joinGroup(final InetAddress mcastaddr) {
    }

    @Override
    public void leaveGroup(final InetAddress mcastaddr) {
    }

    @Override
    public void setNetworkInterface(final NetworkInterface netIf) {
    }

    @Override
    public void send(final DatagramPacket p) {
        mDatagramPacket = p;
    }

    public DatagramPacket getSendPacket() {
        return mDatagramPacket;
    }

    public void setReceiveData(
            final InetAddress address,
            final byte[] data,
            final long wait) {
        mInetAddress = address;
        mReceiveData = data;
        mWait = wait;
    }

    @Override
    public synchronized void receive(final DatagramPacket p) throws IOException {
        if (mReceiveData == null) {
            try {
                Thread.sleep(100000L);
            } catch (final InterruptedException e) {
                throw new IOException();
            }
        }
        try {
            Thread.sleep(mWait);
        } catch (final InterruptedException e) {
            throw new IOException();
        }
        System.arraycopy(mReceiveData, 0, p.getData(), 0, mReceiveData.length);
        p.setLength(mReceiveData.length);
        p.setAddress(mInetAddress);
        mReceiveData = null;
    }
}

File: mmupnp/src/test/java/net/mm2d/upnp/cp/internal/server/MockMulticastSocket.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.cp.internal.server

import java.io.IOException
import java.net.*

class MockMulticastSocket : MulticastSocket() {
    var sendPacket: DatagramPacket? = null
        private set
    private var _inetAddress: InetAddress? = null
    private var _port: Int = 0
    private var receiveData: ByteArray? = null
    private var wait: Long = 0

    @Throws(IOException::class)
    override fun setTimeToLive(ttl: Int) {
    }

    @Throws(IOException::class)
    override fun joinGroup(mcastaddr: InetAddress) {
    }

    @Throws(IOException::class)
    override fun leaveGroup(mcastaddr: InetAddress) {
    }

    @Throws(SocketException::class)
    override fun setNetworkInterface(netIf: NetworkInterface) {
    }

    @Throws(IOException::class)
    override fun send(p: DatagramPacket) {
        sendPacket = p
    }

    fun setReceiveData(address: InetAddress, port: Int, data: ByteArray, wait: Long) {
        _inetAddress = address
        _port = port
        receiveData = data
        this.wait = wait
    }

    @Synchronized
    @Throws(IOException::class)
    override fun receive(p: DatagramPacket) {
        if (receiveData == null) {
            try {
                Thread.sleep(100000L)
            } catch (e: InterruptedException) {
                throw IOException()
            }
        }
        try {
            Thread.sleep(wait)
        } catch (e: InterruptedException) {
            throw IOException()
        }
        val receiveData = receiveData ?: throw IOException()
        System.arraycopy(receiveData, 0, p.data, 0, receiveData.size)
        p.length = receiveData.size
        p.address = _inetAddress
        p.port = _port
        this.receiveData = null
    }
}
